src/scenic/simulators/webots/simulator.py

```diff
@@ -64,34 +74,84 @@ def __init__(self, scene, supervisor, verbosity=0, coordinateSystem=ENU):
         # Find Webots objects corresponding to Scenic objects
         self.webotsObjects = {}
         usedNames = defaultdict(lambda: 0)
+
+        # directory to store proto files for adhoc webots objects
+        tmpMeshDir = tempfile.mkdtemp()
+        # set to store
+        adhocObjectId = 1
+
         for obj in self.objects:
+            # make sure `obj` is a Webots object
             if not hasattr(obj, 'webotsName'):
                 continue    # not a Webots object
-            if obj.webotsName:
-                name = obj.webotsName
+            
+            # get a name of the corresponding webots object
+            name = None
+            if obj.webotsAdhoc:
+                # dynamically generate object from Scenic object mesh
+                if not hasattr(obj.shape, "mesh"):
+                    raise RuntimeError(f'Cannot dynamically instantiate shape without mesh')
+                
+                objFilePath = path.join(tmpMeshDir, f"{adhocObjectId}.obj")
+
+                objectRawMesh = obj.shape.mesh
+                objectScaledMesh = MeshVolumeRegion(mesh=objectRawMesh, dimensions=(obj.width, obj.length, obj.height)).mesh
+
+                trimesh.exchange.export.export_mesh(objectScaledMesh, objFilePath)
+
+                name = f"SCENIC_ADHOC_{adhocObjectId}"
+                
+                rootNode = supervisor.getRoot()
+                rootChildrenField = rootNode.getField("children")
+
+                protoDef = ""
+                if isPhysicsEnabled(obj):
+                    protoDef = dedent(f"""
+                        DEF {name} ScenicObjectWithPhysics {{
+                            url "{objFilePath}"
+                        }}
+                        """
+                    )
+                else:
+                    protoDef = dedent(f"""
+                        DEF {name} ScenicObject {{
+                            url "{objFilePath}"
+                        }}
+                        """
+                    )
```

Comment:
If we're not expecting further differences between these PROTO forms, I would use a single f-string and use another variable to insert `ScenicObject`/`ScenicObjectWithPhysics`.

Reply:
@shumbo Can you take a look at this?